### PR TITLE
Create a variable gangs for grouping By reuse

### PR DESCRIPTION
**Pull request by apmasell**
@mlaszloffy has tested this and seems comfortable with the design.

### CODE REVIEW

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/input/Gang.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package ca.on.oicr.gsi.shesmu.plugin.input;
+
+/** Add a variable to a variable gang */
+public @interface Gang {
```
Comment (callunity) [inline shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/input/Gang.java]: `Add a Shesmu variable to a named group of Shesmu variables`?

From this description it looks like `Gang` refers to a verb (add this to a named group) as opposed to a noun (the named group itself). To me, the verb "gang" conjures images of "gang up on/against", but maybe that's just me? All of which is to say I'd love to sit down with you and talk words.

_pausing review here_

File: implementation.md [hunk 2]
```diff
@@ -167,6 +167,28 @@ the JSON representation and `ttl` indicating the number of minutes to cache the
 input. Additionally, once a Shesmu server is active, it will provide the input
 in the JSON format at `/input/` followed by the input format name.
 
+### Variable Gants
+Variables in an input format can be attached to gang to provide
+convenient grouping criteria. In the `@ShesmuVariable`, set:
+
+    grouping = { @Gang(name = "useful_stuff", order = 0) })
```
[the remark this answers is not in the document]
Comment (apmasell) [reply implementation.md]: Missed in refactoring.

File: language.md [hunk 3]
```diff
@@ -671,6 +671,33 @@ It can also be used in `Let`, `Monitor`, `Run`, and `Alert` clauses in olives:
...
+        By @patient_tissue_prep
+        Into
+          paths = List path
+      Run prep_report With
+        memory = 4Gi,
+        file_name = "{@patient_tissue_prep}"
+        paths = paths;
+
+The gang, in this case, `patient_tissue_prep` defines a number of fields that
+can be grouped together in `By` clause, prefixed with an `@`. The gang can
```
[the remark this answers is not in the document]
Comment (apmasell) [reply language.md]: Not in the current design. Gangs and regular variables exist in separate namespaces, so you could name a variable and a gang the same thing (I'd be very sad if you did though).

File: language.md [hunk 4]
```diff
@@ -671,6 +671,33 @@ It can also be used in `Let`, `Monitor`, `Run`, and `Alert` clauses in olives:
...
+also be converted to a delimited string: `"{@patient_tissue_prep}"`; the order
```
[the remark this answers is not in the document]
Comment (apmasell) [reply language.md]: Right now, it's `_` with no option to change it. I'm open to making it user-definable, but no use case has presently emerged.

File: language.md [hunk 5]
```diff
@@ -887,6 +914,12 @@ determined based on the elements.
 Creates a new named tuple with the fields as specified. The type of the named
 tuple is determined based on the elements.
 
+#### Synthetic Tuple
+- `{@`_name_`}`
+
+Creates a new tuple with the elements as specified in the synthetic group
```
[the remark this answers is not in the document]
Comment (apmasell) [reply language.md]: Bad refactoring. That was the former name for a gang.

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/input/Gang.java (hunk 1 above)
Comment (apmasell) [reply shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/input/Gang.java]: Sure. This name came from a discussion with Morgan and Mike where we looked for synonyms for group.

File: language.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (callunity) [reply language.md]: I don't think there's a need to make it user-definable, but can the delimiter be stated here? "converted to a `_`-delimited string" or similar?

[the remark this answers is not in the document]
Comment (apmasell) [reply language.md]: Fixed.

File: shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/input/Gang.java (hunk 1 above)
Comment (callunity) [reply shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/input/Gang.java]: Let's just go with it.

File: language.md [hunk 6]
```diff
@@ -908,6 +940,7 @@ Specified a new string literal. A string may contain the following special items
 - `{`_expr_`}` for a string interpolation; the expression must be a string, integer, or date
 - `{`_expr_`:`_n_`}` for a zero-padded integer string interpolation; the expression must be an integer and _n_ is the number of digits to pad to
 - `{`_expr_`:`_f_`}` for a formatted date string interpolation; the expression must be a date and _f_ is the [format code](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)
+- `{@`_name_`}` interpolate a name from a gang; the variables in the variables in the gang must be strings and integers
```
Comment (callunity) [inline language.md]: extra `the variables in`

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/TypeUtils.java [hunk 7]
```diff
@@ -1,20 +1,73 @@
...
+                        String.format(
+                            "%d:%d: Cannot find variable ”%s” from gang “%s” as the stream has been manipulated.",
+                            line, column, p.name(), definition.name()));
+                    return Stream.empty();
+                  }
+                  if (!source.get().flavour().isStream()) {
+                    ok.set(false);
+                    errorHandler.accept(
+                        String.format(
+                            "%d:%d: Variable ”%s” from gang “%s” has been shadowed a non-stream local.",
```
Comment (callunity) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/TypeUtils.java]: `shadowed by`

File: shesmu-server/src/test/resources/run/group-synthetic.shesmu [hunk 8]
```diff
@@ -0,0 +1,5 @@
+Input test;
```
Comment (callunity) [inline shesmu-server/src/test/resources/run/group-synthetic.shesmu]: Both of these test files have `synthetic` in the name but I don't actually think that's an issue.

**Pull request merged**